dowhy/causal_refuters/graph_refuter.py

```diff
@@ -31,39 +33,77 @@ def set_refutation_result(self):
+         else:
+            self._false_implications.append([x,y,z])
+            self._results[key]= [cmi_val, False]
       
     def refute_model(self,independence_constraints):
         """
         Method to test conditional independence using the graph refutation object on the given testing set
 
         :param independence_constraints: List of implications to test the conditional independence on
         """
-        if self._method_name is None or self._method_name == "partial_correlation" :
-            refute = GraphRefutation(method_name=self._method_name)
-            for a,b,c in independence_constraints:
-                stats = partial_corr(data=self._data, x= a, y=b, z=list(c))
-                p_value = stats['p-val']
-                key = ((a,b)+(c,))
-                if(p_value < 0.05):
-                    #Reject H0
-                    self._false_implications.append([a,b,c])
-                    self._results[key]= [p_value, False]
+        refute = GraphRefutation(method_name_continuous= self._method_name_continuous ,method_name_discrete = self._method_name_discrete)
+
+        all_nodes = list(self._data.columns.values)
+        discrete_columns = []
+        continuous_columns = []
+        for node in all_nodes:
+            if self._data[node].dtype == np.int64:
+                discrete_columns.append(node)
+            else:
+                continuous_columns.append(node)
+    
+        for a, b, c in independence_constraints:
+            if a in continuous_columns and b in continuous_columns and all(node in continuous_columns for node in c):
+                # a, b and c are all continuous variables
+                if self._method_name_continuous is None or self._method_name_continuous == "partial_correlation":
+                    self.partial_correlation(x = a, y= b, z = c)
                 else:
-                    self._true_implications.append([a, b, c])
-                    self._results[key]= [p_value, True]
-            self.set_refutation_result() #Set refutation result accordingly
-            refute.add_conditional_independence_test_result(number_of_constraints_model = len(independence_constraints), number_of_constraints_satisfied = len(self._true_implications), refutation_result = self._refutation_passed)
-        else:
-            self.logger.error("Invalid conditional independence test")
+                    self.logger.error("Invalid conditional independence test")
+
+            elif a in discrete_columns and b in discrete_columns and all(node in discrete_columns for node in c):
+                # a, b and c are all discrete variables
+                if self._method_name_discrete is None or self._method_name_discrete == "conditional_mutual_information":
+                    self.conditional_mutual_information(x = a, y = b, z =c)
+                else:
+                    self.logger.error("Invalid conditional independence test")
+
+            else:
```

Comment:
This will not work if X or Y is a discrete variable (not just binary). how will you define the partial correlation between X and Y if Y takes three values, say, 'a', 'b', 'c'.
I'd suggest checking for this explicitly. 

1) When Z is discrete, if one of X and Y is continuous and the other is discrete with more than two levels, it is better to discretize X and compute the mutual information.
2) When Z is continuous, if one of X and Y is continuous and the other is discrete with two levels, it is better to consider partial correlation (since binary variable can be represented as {0,1}  values of a single variable)
3) When Z is continuous, if one of X and Y is continuous and the other is discrete with more than two levels, then it is tricky. Perhaps we can ValueError that this setting is not supported.

In the future, complex methods can handle this (e.g., by using kernels), but it is good to establish the different cases explicitly right now.